### ISSUE TITLE
[Bug]: Function call execution does not respect the work_dir

### ISSUE THREAD

**Issue by gagb** [microsoft/autogen#1820]
### Describe the bug

Even if `work_dir` is specified in the code execution config, function call execution does not save files to it.
This prevents building applications that can use function calling properly.

### Steps to reproduce

Run the following code, which I created by register one function in `test/twoagents.py`:
```python
from autogen import AssistantAgent, UserProxyAgent, config_list_from_json

def download_file(url: str, file_path: str, headers: dict = None) -> str:
    """Download a file from a URL and save it to a file path. Return the file path."""
    import requests

    if headers is None:
        headers = {}
    response = requests.get(url, headers=headers)
    with open(file_path, "wb") as file:
        file.write(response.content)
    return file_path

config_list = config_list_from_json(env_or_file="OAI_CONFIG_LIST")
assistant = AssistantAgent("assistant", llm_config={"config_list": config_list})
user_proxy = UserProxyAgent(
    "user_proxy", code_execution_config={"work_dir": "coding", "use_docker": False},
    human_input_mode="NEVER",
    is_termination_msg=lambda msg: msg.get("content") and "TERMINATE" in msg.get("content", "")
)

assistant.register_for_llm(description="Download a file from a URL and save it to a file path.")(download_file)
user_proxy.register_for_execution()(download_file)

user_proxy.initiate_chat(assistant, message="Save the following paper https://arxiv.org/pdf/2308.08155.pdf to disk as wu-arxiv-2023.pdf")
```

It runs and gives the following output

<img width="715" alt="image" src="https://github.com/microsoft/autogen/assets/13227607/4a3c64ec-3c75-4667-88ec-dc19eb19c9b6">

But as a user, I'd expect the file `wu-arxiv-2023.pdf` to be saved in dir called `coding`, but instead it will save it in the cwd as shown below:

<img width="293" alt="image" src="https://github.com/microsoft/autogen/assets/13227607/66f592b2-5dde-492f-b6bf-32c1887f1cfb">

**Comment by gagb** [comment]
@ekzhu @jackgerrits @sonichi

**Comment by gagb** [comment]
It seems that `execute_function` does not even use docker or parameters from `code_execution_config`?

https://github.com/microsoft/autogen/blob/4d0d486115a5445737f7d9937c97ce4453ea15b4/autogen/agentchat/conversable_agent.py#L1959-L2012

**Comment by Noir97** [comment]
Having the same issue

**Comment by ekzhu** [comment]
@Noir97 code execution and function executions are separate, so in function execution it is not going to use the work_dir inside code_execution_config. Take a look at these tutorial chapters.

1. Tool use: https://microsoft.github.io/autogen/docs/tutorial/tool-use
2. Code executor: https://microsoft.github.io/autogen/docs/tutorial/code-executors
3. User defined functions: https://microsoft.github.io/autogen/docs/topics/code-execution/user-defined-functions

**Comment by ekzhu** [comment]
@gagb @jackgerrits perhaps we can close this issue as "won't fix"?

**Comment by gagb** [comment]
I think any func exec should respect constraints set in the executor, but I am not blocked by this in lieu of udfs. So I am fine with this!

**Comment by jtoy** [comment]
I agree with @gagb assessment that it should save to work_dir.  That given can we close this issue since @gagb  is not blocked and it looks like it is something we don't want to fix now.
